### PR TITLE
Creating ADR-0008 -  Choice of test framework for intradomain integration points when testing prior to dev.md

### PR DESCRIPTION
**Pull request by ASQ-EN**
## Description

This ADR will describe the choice of test setup for testing integrations points within a single domain - testing prior to deploying for DEV (U).

## References

* #0000

### CODE REVIEW

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+* ### Databricks integration to a DeltaLake
+
+Write access - storing a time series in the DeltaLake (from Databricks)
+It is assumed that the test will be initiated by placing valid test data on an EventHub from which Databricks will fetch the time series.
+(No seed data required)
+
+* ### Integration to an Sql database
+
+Read/Write access for reading and storing a charge from an azure functions.
+The database must include seed data (MarketParticipants table), Charge test data is not required.
```
Comment (djorgensendk) [inline docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:55]:
```suggestion
Read/Write access for reading and storing a record from an application.
The database must be seeded with appropiate test data prior to running the application code.
```

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+***Regarding the use case for DeltaLake - it has not yet been finalized on which solution is most valuable, and therefore the decision to cover this is still pending.***
+
+### Positive Consequences <!-- optional -->
+
+* Besides testing component connections locally Squadron allows verification that components are connecting correctly to storage instances by creating a containerized setup where both application layer and service is connected in an infrastructure setup (excluding Terraform - further outline on this under negative consequences).
+
+* The solution is implemented to run both locally and in CI setup.
+
+* This adds to  even more risk elements to be covered prior to deployment, and mitigiates risks that essentially could cause instability and issues on environments and further out in the CD pipeline, which would enhance the complete lead time for functionality delivery.
```
Comment (djorgensendk) [inline docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:77]: This sounds like a general positive consequence of implementing tests where one verify connections and data exchanges with external dependencies - and not a positive consequence specifically with the Squadron/containernized approach

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+### Negative Consequences <!-- optional -->
+
+One risk area that has been mentioned for this strategy for testing prior to dev is that the solution does not cover testing of compliance and quality of the solution deployed with correct Terraform modules. This risk is also be present in a test container setup - but could be mitigated by a PR environment solution. We are already using a PR environment solution to ensure test on Terraform and infrastructure - and it was also tested as a solution for intra-domain integration testing. This is can be done, but depending on the testing discipline a very expensive solution, and is therefore not the approach we have choosed.
```
Comment (djorgensendk) [inline docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:81]:
```suggestion
One risk area that has been mentioned for this strategy for testing prior to dev is that the solution does not cover testing of compliance and quality of the solution deployed with correct Terraform modules. This risk is also be present in a test container setup - but could be mitigated by a PR environment solution. We are already using a PR environment solution to ensure that the Terraform based Infrastructure as Code works as expected. The PR environment was also tested as a solution for intra-domain integration testing. This is can be done, but depending on the testing discipline a very expensive solution, and is therefore not the approach we have choosed.
```

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+Given the isolated testing on components and infrastructure, we are mitigating the risk of this issue, by having the possibility of testing services in the PR environment - but given price plans etc the PR environment is triggered in PR gate and torn down immedialtely after given tests have been performed, and should not be used for debugging in the PR gate over a longer period of time.
```
Comment (djorgensendk) [inline docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:83]: Of course it is possible to run tests on the PR environment, but I believe we must completely get away from this approach? Anyways if the PR environment is torn down immediately testing on it won't be possible

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+* We emphasize failing fast, having quick and reliable feedback loops and DEV environment stability and availability
+* We want a solution that does not require a full domain setup in an environment
+* The solution must be easy to set up and tear down as part of PR gate or deployment gates
+* We want a solution that are not affected by infrastructure, so test results are not biased by infrastructure issues
+
+Following use cases were identified, that needs to be covered by this test framework:
+
+* ### Triggering job in Ccomponents
+
+Being able to trigger, read status and verify output from the component.
```
[the remark this answers is not in the document]
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:34]: But if we delete it, then the use case is not ensured, are we sure that we are not leaving Joules with a need uncovered?

[the remark this answers is not in the document]
Comment (djorgensendk) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:34]: In its currect form I believe it should be deleted. Then we must talk to Joules if they need an example for how to communicate with Blob storage. And if that's the case, then write a new section focusing on Blob storage similar to the other sections in the ADR

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD (hunk 4 above)
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:81]: could we agree that line 83 is then removed totally?

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD (hunk 2 above)
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:75]: Will be moved to a more general topic on why we do this

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+To be able to ensure that our integrations componenets are set up correctly and functional with expected behavior before deploying to DEV (U). We need to use a framework that allows us to create local instances for testing prior to deployment to DEV environment.
+
+## Decision Drivers
+
+* We need an integration test solution that fits a microservice architecture as well as the practices of DevOps
+* The solution must solve the problem with cloud components that cannot be emulated, specifically for local testing efforts
+* We emphasize failing fast, having quick and reliable feedback loops and DEV environment stability and availability
+* We want a solution that does not require a full domain setup in an environment
+* The solution must be easy to set up and tear down as part of PR gate or deployment gates
```
[the remark this answers is not in the document]
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:27]: Removed

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+* We want a solution that are not affected by infrastructure, so test results are not biased by infrastructure issues
```
[the remark this answers is not in the document]
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:28]: have added explanation to bullet four and removed bullet 6

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
```
[the remark this answers is not in the document]
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:29]: added

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD (hunk 5 above)
[the remark this answers is not in the document]
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:34]: Removed - must be referenced to test strategy that azure components functionality are tested elsewhere

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+Publishing a ChargeCommandReceived event from an azure function (MessageReceiver). Receiving this event from yet another az function (ChargeCommandReceiver).
+The test will be initiated by a HTTP POST call with JSON content.
+(No seed data required)
+
+* ### Databricks integration to a DeltaLake
+
+Write access - storing a time series in the DeltaLake (from Databricks)
+It is assumed that the test will be initiated by placing valid test data on an EventHub from which Databricks will fetch the time series.
+(No seed data required)
```
[the remark this answers is not in the document]
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:48]: Done

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD (hunk 4 above)
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:81]: Done

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+* ### Integration to Event Hub
+
+Verifying event is generated in event hub and that it contains expected data.
+
+* ### Integration to a ServiceBus
+
+Publishing a ChargeCommandReceived event from an azure function (MessageReceiver). Receiving this event from yet another az function (ChargeCommandReceiver).
+The test will be initiated by a HTTP POST call with JSON content.
+(No seed data required)
```
[the remark this answers is not in the document]
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD]: Done

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD (hunk 9 above)
[the remark this answers is not in the document]
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD]: description changed to above forementioned

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD (hunk 4 above)
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD]: removed

File: docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD (hunk 8 above)
[the remark this answers is not in the document]
Comment (ASQ-EN) [reply docs/architecture-decision-record/ADR-0008 - Choice of test framework for intradomain integration points when testing prior to dev.MD:29]: Done

**Pull request closed without merging**